subsys/net/ip/tcp.c

```diff
@@ -1092,19 +1097,23 @@ static int tcp_send_queued_data(struct tcp *conn, size_t new_packet_size)
 	}
 
 	while (tcp_unsent_len(conn) > 0) {
-		if (tcp_window_full(conn)) {
-			subscribe = true;
-			break;
-		}
-
 		ret = tcp_send_data(conn);
 		if (ret < 0) {
 			break;
 		}
 	}
 
 	if (tcp_window_full(conn)) {
-		(void)k_sem_take(&conn->tx_sem, K_NO_WAIT);
+		/* When new data is transmitted it will be removed from the list
+		 * afterwards, do not take the semaphore
```

Comment:
Then I think this function should be moved out of this function, and window should be verified from within `net_tcp_queue_data()`/`tcp_in()` separately.

Reply:
Makes a lot of sense, this simplifies the is logic a ton, will fix it